server/repository/src/db_diesel/changelog/changelog.rs

```diff
@@ -0,0 +1,199 @@
+    pub store_id: Option<EqualFilter<String>>,
+    pub record_id: Option<EqualFilter<String>>,
+}
+
+pub struct ChangelogRepository<'a> {
+    connection: &'a StorageConnection,
+}
+
+impl<'a> ChangelogRepository<'a> {
+    pub fn new(connection: &'a StorageConnection) -> Self {
+        ChangelogRepository { connection }
+    }
+
+    /// Returns changelog rows order by operation sequence in asc order
+    ///
+    /// # Arguments
+    ///
+    /// * `earliest` - Starting cursor (first returned changelogs may be ahead in sequence from starting cursor)
+    /// * `limit` - Maximum number of entries to be returned
+    /// * `filter` - Extra filter to apply on change_logs
+    pub fn changelogs(
+        &self,
+        earliest: u64,
+        limit: u32,
+        filter: Option<ChangelogFilter>,
+    ) -> Result<Vec<ChangelogRow>, RepositoryError> {
+        let query = create_filtered_query(earliest, filter).limit(limit.into());
+
+        // // Debug diesel query
+        // println!(
+        //     "{}",
+        //     diesel::debug_query::<crate::DBType, _>(&query).to_string()
+        // );
+
+        let result = query.load(&self.connection.connection)?;
+        Ok(result)
+    }
+
+    pub fn count(
+        &self,
+        earliest: u64,
+        filter: Option<ChangelogFilter>,
+    ) -> Result<u64, RepositoryError> {
+        let result = create_filtered_query(earliest, filter)
+            .count()
+            .get_result::<i64>(&self.connection.connection)?;
+        Ok(result as u64)
+    }
+
+    // TODO we only really want to know the cursor, so just max ?
+    pub fn latest_changelog(&self) -> Result<u64, RepositoryError> {
+        let result = changelog::dsl::changelog
+            .select(diesel::dsl::max(changelog::dsl::cursor))
+            .first::<Option<i64>>(&self.connection.connection)?;
+        Ok(result.unwrap_or(0) as u64)
+    }
+}
+
+type BoxedChangelogQuery = IntoBoxed<'static, changelog_deduped::table, DBType>;
+
+fn create_filtered_query<'a>(
+    earliest: u64,
+    filter: Option<ChangelogFilter>,
+) -> BoxedChangelogQuery {
+    let mut query = changelog_deduped::dsl::changelog_deduped
+        .filter(changelog_deduped::dsl::cursor.ge(earliest.try_into().unwrap_or(0)))
+        .into_boxed();
+
+    if let Some(f) = filter {
+        let ChangelogFilter {
+            table_name,
+            name_id,
+            store_id,
+            record_id,
+        } = f;
+
+        apply_equal_filter!(query, table_name, changelog_deduped::dsl::table_name);
+        apply_equal_filter!(query, name_id, changelog_deduped::dsl::name_id);
+        apply_equal_filter!(query, store_id, changelog_deduped::dsl::store_id);
+        apply_equal_filter!(query, record_id, changelog_deduped::dsl::record_id);
+    }
+
+    query
+}
+
+impl Default for ChangelogRow {
+    fn default() -> Self {
+        Self {
+            row_action: ChangelogAction::Upsert,
+            table_name: ChangelogTableName::Invoice,
```

Comment:
Feels a bit weird defaulting to this. I assume this is just used for tests?

Reply:
Sure, added a cfg flag: d29d57db47f68f7237714d20f3796ba714e34511, will be checked/tested in the `head` pr before merging

Reply:
Didn't work, just added comment in the `head` branch: 97a8c00e6418e88ea6ae6b6a8ccbff918f068d77